### DISCUSSION TITLE
Inquiry Regarding Automated Termination of Simulation Based on Runtime Threshold

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#27468] [Q&A General] [answered]
Dear Moose Forum Members,

I hope this message finds you well. I am currently seeking assistance regarding the automatic termination of simulations when the runtime exceeds a specified threshold.

I have been utilizing `PerfGraphData` to acquire runtime information and subsequently employing `Terminator` to set termination conditions in my input file, as demonstrated below:

```bash
[UserObjects]
  [./term]
    type = Terminator
    expression = 'run_time > 600'
  [../]
[]
[Postprocessors]
...
  [./run_time]
    type = PerfGraphData
    section_name = "Root"
    data_type = total
  [../]
...
[./]
```

However, I have encountered an issue where `Terminator` and `PerfGraphData` only execute at the beginning or end of a timestep, rather than terminating the simulation once the overall runtime exceeds a certain threshold. Therefore, my inquiry is whether there exists a method or condition to automatically terminate the simulation when it reaches a specified runtime threshold, such as after 10 minutes of computation.

The reason behind my quest for this functionality is that in the simulations I have created, the output continuously displays "Computing Residual," as illustrated below:

```
Time Step 8, time = 0.108816, dt = 0.020736

Grain Tracker Status:
Grains active index 0: 1 -> 1
Grains active index 1: 1 -> 1

Finished inside of GrainTracker

    Computing Residual............................................................................................................................................
```

However, attempts to terminate the program using `ctrl + c` or `killall` have been unsuccessful in generating *.prof files. On the contrary, the automatic termination of the simulation has successfully generated *.prof files.

I would greatly appreciate any insights or suggestions on how to achieve automated termination based on a runtime threshold.

Thank you for your time and assistance.

Best regards,

Wei

**Comment by GiudGiud** [comment, marked as answer]
Hello

The default execution schedule for these objects is at the end of every time step

You ll want to specify execute_on of LINEAR to have them execute much more often. Then they can stop the simulation at any linear iteration (when computing the residual in reality)

Guillaume

**Reply by PengWei97** [reply]
Hi @GiudGiud,

Now I followed your suggestion of doing it during the linear step, but it still doesn't work, as it turns out that our challenge is that when the simulation gets stuck at "Compute Residuals............", we can't do the linear step ends.

To solve this problem, I think it is crucial to ensure that the program gracefully catches the interrupt signal and performs the necessary cleanup and save operations before exiting. However, this seems to involve modifying the main() function.

I will keep you updated on any developments. If you have any insights I'd be grateful for your input.

Sincerely,
Wei

**Reply by GiudGiud** [reply]
The interrupt signal is gracefully caught already, it s just that the Terminator is only emitting it when it is executed

And seemingly it does not get executed. Or not yet.

**Reply by PengWei97** [reply]
Yes, the Terminator is only executed when it evaluates the expression = 'run_time > 600', so the Terminator cannot be called when the simulation gets stuck at 'Compute Residuals............'. 

Therefore, can the simulation program perform the necessary cleanup and save operations before exiting when Ctrl+C is executed?

By the way, if I set execute_on=CUSTOM, can I customize the behavior of the terminator to trigger when the real-time running time of the simulation reaches a certain threshold?

**Reply by GiudGiud** [reply]
`execute_on=CUSTOM` will still require a place somewhere in the code where you will put `_problem.execute(EXEC_CUSTOM)`. So unless you know a good place within computeResidual (but I expect it's taking long because the mesh is large and simply computeQpResidual on every kernel is taking a lot of time?), it won't work

> Therefore, can the simulation program perform the necessary cleanup and save operations before exiting when Ctrl+C is executed?

No we don't have code to catch SIGINT or SIGSTP in moose. So we would not properly stop the simulation.
However I dont know that you need this to profile? Did you try? Maybe you can profile even without a clean program termination

**Reply by PengWei97** [reply]
Hi @GiudGiud,

Thank you for your efforts. This problem has been solved by setting `signal_received` to accept the `Ctrl + C` signal and the execute mooseError. The simulation can then perform the operation of saving the performance report file.

Sincerely,
Wei
